### PR TITLE
Add ASV benchmark CI workflow

### PR DESCRIPTION
**Pull request by Illviljan**
Add an ASV workflow for easy benchmarking. Adapted from xarray.

### CODE REVIEW

**Comment by Illviljan** [comment]
@dcherian, could you add the `run-benchmark` label?

**Comment by dcherian** [comment]
You have a pending invite to xarray-contrib, if you accept I think you have rights to do that. (I'd like to make sure you can do it so not doing it right now)

**Comment by Illviljan** [comment]
Thanks, it works!

**Comment by Illviljan** [comment]
The workflow appears to work!

But the benchmark is probably broken?
```
 Error:                Traceback (most recent call last):
                 File "/home/runner/.local/lib/python3.8/site-packages/asv/benchmark.py", line 1293, in main_run_server
                   main_run(run_args)
                 File "/home/runner/.local/lib/python3.8/site-packages/asv/benchmark.py", line 1167, in main_run
                   result = benchmark.do_run()
                 File "/home/runner/.local/lib/python3.8/site-packages/asv/benchmark.py", line 573, in do_run
                   return self.run(*self._current_params)
                 File "/home/runner/.local/lib/python3.8/site-packages/asv/benchmark.py", line 669, in run
                   samples, number = self.benchmark_timing(timer, min_repeat, max_repeat,
                 File "/home/runner/.local/lib/python3.8/site-packages/asv/benchmark.py", line 705, in benchmark_timing
                   timing = timer.timeit(number)
                 File "/home/runner/work/flox/flox/asv_bench/.asv/env/b75ad52f6dea6a19ad824a655424eb23/lib/python3.9/timeit.py", line 177, in timeit
                   timing = self.inner(it, self.timer)
                 File "<timeit-src>", line 6, in inner
                 File "/home/runner/.local/lib/python3.8/site-packages/asv/benchmark.py", line 627, in <lambda>
                   func = lambda: self.func(*param)
                 File "/home/runner/work/flox/flox/asv_bench/benchmarks/combine.py", line 16, in time_combine
                   flox.core._npg_combine(
               AttributeError: module 'flox.core' has no attribute '_npg_combine'
               asv: benchmark failed (exit status 1)
```

Maybe we should merge this and make a new PR fixing the benchmarks?

**Comment by dcherian** [comment]
Could just fix it here.

Is it necessary to change these paths? If so, I think you need to also add a change deleting the existing file.

**Comment by Illviljan** [comment]
Do you see what needs to be changed? I'm not familiar enough with this code and I can't find something that resembles the old  `flox.core._npg_combine`?


The paths were changed to align with xarray and I don't want to fiddle with the paths. I still have nightmares of getting the correct paths when adding this in xarray.

From what I can tell it seems to understand that it was moved or?
image

File: asv_bench/benchmarks/combine.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+N = 1000
+
+
+class Combine:
+    def setup(self, *args, **kwargs):
+        raise NotImplementedError
+
+    @parameterized("kind", ("cohorts", "mapreduce"))
+    def time_combine(self, kind):
+        flox.core._npg_combine(
```
Comment (dcherian) [inline asv_bench/benchmarks/combine.py:16]:
```suggestion
        flox.core._grouped_combine(
```

File: asv_bench/benchmarks/combine.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            getattr(self, f"x_chunk_{kind}"),
+            **self.kwargs,
+            keepdims=True,
+            engine="numpy",
+        )
+
+    @parameterized("kind", ("cohorts", "mapreduce"))
+    def peakmem_combine(self, kind):
+        flox.core._npg_combine(
```
Comment (dcherian) [inline asv_bench/benchmarks/combine.py:25]:
```suggestion
        flox.core._grouped_combine(
```

**Review by dcherian — COMMENTED** [review]
I mean that the `benchmarks` folder has been moved but in the diff it doesn't show this. It looks like `asv_bench/benchmarks/combine.py` was added, but the earlier `benchmarks/combine.py` was not removed

File: asv_bench/benchmarks/combine.py [hunk 3]
```diff
@@ -58,4 +58,4 @@ def construct_member(groups):
         ]
 
         self.x_chunk_cohorts = [construct_member(groups) for groups in [np.array((1, 2, 3, 4))] * 4]
-        self.kwargs = {"agg": flox.aggregations.mean, "axis": (3,), "neg_axis": (-1,)}
+        self.kwargs = {"agg": flox.aggregations.mean, "axis": (3,), "group_ndim": 1}
```
Comment (dcherian) [inline asv_bench/benchmarks/combine.py:61]:
```suggestion
        self.kwargs = {"agg": flox.aggregations.mean, "axis": (3,), "neg_axis": (-1,)}
```

File: asv_bench/benchmarks/reduce.py [hunk 4]
```diff
@@ -1,44 +1,33 @@
 import numpy as np
-import numpy_groupies as npg
```
Comment (dcherian) [inline asv_bench/benchmarks/reduce.py:2]: I think this is wrong and you've copied an old version of the file.

File: .github/workflows/benchmarks.yml [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+      # https://github.com/actions/checkout/issues/23
+      - uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+
+      - name: Setup Miniconda
+        uses: conda-incubator/setup-miniconda@v2
+        with:
+          # installer-url: https://github.com/conda-forge/miniforge/releases/latest/download/Mambaforge-Linux-x86_64.sh
+          installer-url: https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-Linux-x86_64.sh
```
Comment (dcherian) [inline .github/workflows/benchmarks.yml:27]: Could micromamba this like with the other workflows and get some caching.

File: asv_bench/benchmarks/reduce.py (hunk 4 above)
Comment (Illviljan) [reply asv_bench/benchmarks/reduce.py:2]: Hmm, might've had an old fork...

File: .github/workflows/benchmarks.yml [hunk 6]
```diff
@@ -0,0 +1,74 @@
+name: Benchmark
+
+on:
+  pull_request:
+    types: [opened, reopened, synchronize, labeled]
+  workflow_dispatch:
+
+jobs:
+  benchmark:
+    if: ${{ contains( github.event.pull_request.labels.*.name, 'run-benchmark') && github.event_name == 'pull_request' || github.event_name == 'workflow_dispatch' }}
```
Comment (dcherian) [inline .github/workflows/benchmarks.yml:10]:
```suggestion
    if: ${{ github.event_name == 'pull_request' || github.event_name == 'workflow_dispatch' }}
```

this is running in 3 minutes; we can just run it always.

Comment (Illviljan) [reply .github/workflows/benchmarks.yml]: Done.

File: .github/workflows/benchmarks.yml [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+          environment-name: flox-tests
+          cache-env: true
+          # extra-specs: |
+          #   python="${{ matrix.python-version }}"
+
+      - name: Setup some dependencies
+        shell: bash -l {0}
+        run: |
+          pip install asv
+          sudo apt-get update -y
```
Comment (Illviljan) [inline .github/workflows/benchmarks.yml:37]: Can this be avoided as well? if we just add asv to ci/environment.yml?

File: .github/workflows/benchmarks.yml (hunk 5 above)
Comment (Illviljan) [reply .github/workflows/benchmarks.yml]: I copied from the other worklows and it seems to be working? Does this make sense to?

File: .github/workflows/benchmarks.yml (hunk 7 above)
Comment (Illviljan) [reply .github/workflows/benchmarks.yml]: Appears so!

**Review by dcherian — APPROVED** [review]
Looks great. thanks @Illviljan feel free to merge when you're done.

**Pull request merged**